### PR TITLE
[metrics] Support macro averaging.

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Bring back support for macro averaging, which was lost in the metrics.

After discussion during the team meeting, I continued to look at the code more and more and determined that I still did not like the proposed solution of mismatching metrics during train/eval time (which was one of the major motivations of user level metrics). As such, I've implemented proper macro averaging as a higher level metric.

The only part I'm really unhappy about is carving out the exception for SumMetric. It's a bit annoying because `exs` is a critical metric that must be reported, and we don't want that averaged per-task, we want it summed.

**Testing steps**
Manual testing. New CI. Long CI of older things. @klshuster  can we run a test with your dodeca models (just point me to the model files in chat and a task mixture)

~Running locally, I think there might be a hang around when using with multiprocessing train.~ Resolved.

### CODE REVIEW

File: tests/test_metrics.py [hunk 1]
```diff
@@ -89,10 +98,12 @@ def test_average_metric_additions(self):
             self.assertAlmostEqual(actual_output, output, places=6)
             self.assertIsInstance(actual_output, float)
 
-    def test_fixedmetric(self):
```
Comment (EricMichaelSmith) [inline tests/test_metrics.py]: It's not still useful to test FixedMetric here?

File: tests/test_metrics.py [hunk 2]
```diff
@@ -173,5 +184,82 @@ def test_largebuffer(self):
...
+            'fixed': FixedMetric(4),
+            'global_avg': GlobalAverageMetric(3, 4),
+        }
+        report2 = {
+            'avg': AverageMetric(1, 3),
+            'sum': SumMetric(4),
+            'fixed': FixedMetric(4),
+            'global_avg': GlobalAverageMetric(1, 3),
+        }
+        agg = aggregate_unnamed_reports([report1, report2])
```
Comment (EricMichaelSmith) [inline tests/test_metrics.py]: Hmm I suppose we can safely assume that all unnamed reports will never need to be macroaveraged?

File: parlai/core/metrics.py [hunk 3]
```diff
@@ -411,26 +444,43 @@ def normalize_answer(s):
...
                     m[each_metric] = value
             else:
-                # none + a => a from implementation of Metric.__add__
-                m[each_metric] = m.get(each_metric) + value
-                if len(named_reports) > 1:
-                    task_metric = f'{task_id}/{each_metric}'
-                    m[task_metric] = m.get(task_metric) + value
+                task_metric = f'{task_id}/{each_metric}'
+                m[task_metric] = m.get(task_metric) + value
+                if micro_average or isinstance(value, (SumMetric, FixedMetric)):
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Ah, here's the `SumMetric` carve-out. Since we're implementing macro-*averaging* here, why don't we have the condition be that you only macroaverage if it's an `AverageMetric`?

File: tests/test_metrics.py (hunk 2 above)
Comment (stephenroller) [reply tests/test_metrics.py]: We can implement it later as we need. For now, unnamed reports is for two cases:
- Aggregating across workers
- Aggregating across batches

Both cases will always want micro averaging.

File: parlai/core/metrics.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: Yeah I wrestled with that. PPL stands out as a weird edge case. It is an average metric, so it would work. What does @klshuster think?

File: tests/test_metrics.py (hunk 1 above)
Comment (stephenroller) [reply tests/test_metrics.py]: You're right, I should leave it.

File: parlai/core/metrics.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: Updated per your suggestion.

File: parlai/core/metrics.py [hunk 4]
```diff
@@ -231,9 +231,42 @@ def value(self) -> float:
...
+
+
 class GlobalMetric:
     """
     A global metric is one that should not be aggregated across different tasks.
+
+    Key to it is the notion that any one worker or any one task already has a global
+    view of the value, and so no combinations should be done. Note this is different
+    then a FixedMetric, in that a GlobalMetric can be still averaged across multiple
+    parleys(), but a FixedMetric is always fixed.
```
Comment (emilydinan) [inline parlai/core/metrics.py]: can you give examples of each of these metrics in the descriptiosn? I find it slightly confusing

File: parlai/core/metrics.py [hunk 5]
```diff
@@ -411,26 +444,43 @@ def normalize_answer(s):
...
 
-def aggregate_named_reports(named_reports: Dict[str, Dict[str, Metric]]):
+def aggregate_named_reports(
+    named_reports: Dict[str, Dict[str, Metric]], micro_average: bool = False
+):
     """
     Aggregate metrics from multiple reports.
 
     :param reports: Dict of tasks -> metrics.
+    :param micro_average: If true, top level metrics will be the macro average.
```
Comment (klshuster) [inline parlai/core/metrics.py]: nit: `If true, top level metrics will be the _micro_ average`

File: parlai/nn/lr_scheduler.py [hunk 6]
```diff
@@ -91,6 +92,7 @@ def load_state(self, states):
         if states.get('warmup_scheduler') and getattr(self, 'warmup_scheduler', None):
             self.warmup_scheduler.load_state_dict(states['warmup_scheduler'])
         self._number_training_updates = states.get('number_training_updates', 0)
+        self.step(self._number_training_updates)
```
Comment (klshuster) [inline parlai/nn/lr_scheduler.py]: what is this change?

File: parlai/scripts/train_model.py [hunk 7]
```diff
@@ -182,7 +182,16 @@ def setup_args(parser=None) -> ParlaiParser:
         'ppl,f1,accuracy,hits@1,rouge,bleu'
         'the rouge metrics will be computed as rouge-1, rouge-2 and rouge-l',
     )
+    train.add_argument(
```
Comment (klshuster) [inline parlai/scripts/train_model.py]: perhaps a totally different issue here but seeing this param added twice leads me to believe we should have some sort of `common_script_args` that we can add for all our scripts (would help standardize command line params e.g. how `-n` is num examples in some while `-ne` is num examples in others)

File: parlai/core/metrics.py (hunk 5 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: hardly a nit!

File: parlai/nn/lr_scheduler.py (hunk 6 above)
Comment (stephenroller) [reply parlai/nn/lr_scheduler.py]: oh boy, it's a bit of a pytorch hack.

Until we do a step(), the LR from a FRESHLY loaded model may report an incorrect value. The changes in train model that force a reload surfaced this.

File: parlai/core/metrics.py (hunk 4 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: Done for all the metrics classes.
